oxyno-zeta/s3-proxy
# feat: Add TLS (SSL) support.

---
name: Add TLS (SSL) support.
about: This adds TLS (SSL) support for s3-proxy, allowing us to encrypt the link between the load balancer and the S3 Proxy instances.
---

## Issue/Feature

This enables both the server and internal server to encrypt data using Go's builtin TLS routines from crypto/tls.

## Additional Information

Compliance folks are increasingly scrutinizing the link between the load balancer and servers in AWS, with some deeming security groups and NACLs insufficient controls and demanding this link be encrypted.

This enables the link to be encrypted using either generated, self-signed certificates/private keys or certificates and private keys specified directly, in S3, in AWS Secrets Manager, or AWS Systems Manager (SSM) Parameter Store.

Note that AWS Certificate Manager is not supported because it does not allow you to obtain the private key (by design).

## Verification Steps

Add a new section to `server` or `internalServer` in your config named `ssl`:

```yaml
server:
    ssl:
        enabled: true # Optional; defaults to true if certificates/selfSignedHostnames is set, false otherwise.
        certificates:
            -  # Exactly one of certificate/certificateUrl and privateKey/privateKeyUrl must be specified.
                certificate: |
                    -----BEGIN CERTIFICATE-----
                    ....
                    -----END CERTIFICATE-----
                certificateUrl: arn:aws:ssm:region:accountId:parameter/name
                privateKey: |
                    -----BEGIN RSA PRIVATE KEY-----
                    -----END RSA PRIVATE KEY-----
                privateKeyUrl: arn:aws:secretsmanager:region:accountId:secret/name
        selfSignedHostnames:
            - localhost
            - localhost.localdomain
        minTLSVersion: TLSv1.2  # Defaults to TLSv1.2
        maxTLSVersion: TLSv1.3  # Defaults to no limit
        cipherSuites: # See https://pkg.go.dev/crypto/tls#pkg-constants for valid names; the current default is shown below.
            - TLS_AES_128_GCM_SHA256
            - TLS_AES_256_GCM_SHA384
            - TLS_CHACHA20_POLY1305_SHA256
            - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
            - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
            - TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
```

## Checklist:
